File: src/numi.py
import json
from .utils.file_io import Content
from .parser.lexer import Lexer
from .parser.parser import Parser
from .compiler.compiler import Compiler

class Numi:
    def __init__(self, path):
        self.path: str = path
        
        # Content
        self.content = Content(self.path)
        self.source  = self.content.get_content()

        # Lexer
        self.lexer  = Lexer(self.path, self.source)
        self.tokens = self.lexer.lex()
        
        # Parser
        self.parser = Parser(self.tokens, self.path)
        self.ast    = self.parser.generateAST()
        
        # Compiler
        self.compiler = Compiler(self._file())
    
    def _file(self):
        # handle /
        _ret = ""
        if "/" in self.path:
            _ret = self.path.split("/")[-1]
        elif "\\" in self.path:
        # handle \
            _ret = self.path.split("\\")[-1]

        if "." in _ret:
                _ret = _ret.rsplit(".", 1)[0]

        return _ret

    def get_ast(self):
        return json.dumps(self.ast, indent=4)

    def compile(self):
        self.compiler.compile()

File: src/numi.js
const { Content } = require("./utils/fileIO")
const { Lexer } = require("./parser/lexer/lexer")
const { Parser } = require("./parser/parser")
const { Compiler } = require("./compiler/compiler")
const path = require("path")

class Numi {
    constructor(filePath) {
        this.path = filePath
        
        // Content
        this.content = new Content(this.path)
        this.source = this.content.get_content()

        // Lexer
        this.lexer  = new Lexer(this.path, this.source)
        this.tokens = this.lexer.lex()
        this.tokens_cpy = this.tokens.slice()
        
        // Parser
        this.parser = new Parser(this.path, this.tokens_cpy)
        this.ast    = this.parser.generateAST()
        
        // Compiler
        this.compiler = new Compiler(this.path, path.parse(this.path).name)
    }

    dump_tokens() {
        return console.log(this.tokens)
    }

    dump_ast() {
        return console.log(JSON.stringify(this.ast, null, 4))
    }

    compile() {
        return console.log(this.compiler.compile(this.ast)) // temp
    }

}

module.exports = {
    Numi
}

if (require.main === module) {
    const process = require("process")

    console.error(`ERROR: \`${path.parse(__filename).name}\` file can\'t be ran as main.`)
    process.exit(1)
}